Utils/Types/AtomicQueue.hpp

```diff
@@ -0,0 +1,90 @@
+// ======================================================================
+// \title  AtomicQueue.hpp
+// \brief  Lightweight lock-free non-allocating single reader single writer queue
+// ======================================================================
+
+#ifndef UTILS_TYPES_ATOMIC_QUEUE_HPP
+#define UTILS_TYPES_ATOMIC_QUEUE_HPP
+
+#include <atomic>
+
+namespace Types {
+
+// Note: PlatformSizeType is probably generally larger than we need,
+// but it should still be an efficient size to manipulate,
+// and it's guaranteed to be unsigned, which is crucial.
+template<class E, PlatformSizeType SIZE>
+class AtomicQueue {
+public:
+    static_assert(SIZE * 2 <= std::numeric_limits<PlatformSizeType>::max(),
+        "This implementation distinguishes full and empty queues by using indices modulo SIZE * 2,"
+        "so SIZE * 2 must fit in the index type");
+
+    bool isFull() {
+        return countElements(m_nextEnqueueIdx.load(), m_nextDequeueIdx.load()) == SIZE;
+    }
+
+    bool isEmpty() {
+        return countElements(m_nextEnqueueIdx.load(), m_nextDequeueIdx.load()) == 0;
+    }
+
+    // Single writer
+    bool enqueue(const E &element) {
+        PlatformSizeType nextDequeueIdx = m_nextDequeueIdx.load();
+        PlatformSizeType nextEnqueueIdx = m_nextEnqueueIdx.load();
+
+        if (countElements(nextEnqueueIdx, nextDequeueIdx) == SIZE) {
+            return false;
+        }
+
+        m_elements[nextEnqueueIdx % SIZE] = element;
+        m_nextEnqueueIdx.store((nextEnqueueIdx + 1) % (SIZE * 2));
+        return true;
+    }
+
+    // Single reader
+    bool dequeue(E &elementOut) {
+        PlatformSizeType nextDequeueIdx = m_nextDequeueIdx.load();
+        PlatformSizeType nextEnqueueIdx = m_nextEnqueueIdx.load();
+
+        if (countElements(nextEnqueueIdx, nextDequeueIdx) == 0) {
+            return false;
+        }
+
+        elementOut = m_elements[nextDequeueIdx % SIZE];
+        m_nextDequeueIdx.store((nextDequeueIdx + 1) % (SIZE * 2));
+        return true;
+    }
+
+    bool peek(E &elementOut) {
+        PlatformSizeType nextDequeueIdx = m_nextDequeueIdx.load();
+        PlatformSizeType nextEnqueueIdx = m_nextEnqueueIdx.load();
+
+        if (countElements(nextEnqueueIdx, nextDequeueIdx) == 0) {
+            return false;
+        }
+
+        elementOut = m_elements[nextDequeueIdx % SIZE];
+        return true;
+    }
+
+    bool dequeue() {
+        E ignored;
+        return dequeue(ignored);
+    }
+
+private:
+    E m_elements[SIZE];
+    std::atomic<PlatformSizeType> m_nextEnqueueIdx;
```

Comment:
These variables should be initialized to zero, either here or in the constructor initializers.